### PR TITLE
Feature/resolve css declarations

### PR DESCRIPTION
**Pull request by tylerferguson**

### CODE REVIEW

**Comment by tylerferguson** [comment]
Hello! Thought I could create this for some discussion around #15 . This is one way we could do this if there was appetite for such functionality.

File: src/serializers/styleSheetSerializer.js [hunk 1]
```diff
@@ -38,8 +38,39 @@ const getMediaQueries = (ast, filter) => (
...
+    .reverse()
+    .filter(declaration => {
+      if (declarationsMap[declaration.property]) {
+        return false
+      }
+
+      declarationsMap[declaration.property] = true
+      return true
+    })
+    .reverse()
```
Comment (tylerferguson) [inline src/serializers/styleSheetSerializer.js]: The two reverses are a tad novel in this function and could be done with a simple reverse loop. Sometimes I like novelty :D

**Comment by MicheleBertoli** [comment]
Thanks @tylerferguson for implementing this.

Your implementation is simple (and I like it) but, as far as I understand, your goal is adding to the snapshot only the properties that you actually see in the browser, right?

In that case, there are a few edge cases which are not covered:

**!important**

```js
const Button = styled.button`
  background: blue !important;
  border-radius: 3px;
  border: none;
  color: white;
`

const TomatoButton = styled(Button)`
  background: tomato;
`
```

**-vendor-prefixes**

```js
const Button = styled.button`
  background: palevioletred;
  border-radius: 3px;
  border: none;
  color: white;
`

const TomatoButton = styled(Button)`
  background: tomato;
  -webkit-border-radius: 0;
`
```

Again, I believe snapshot should represent the output and I'm still not sure this is the right way but I'm more than happy to hear other people opinions.

File: src/serializers/styleSheetSerializer.js [hunk 2]
```diff
@@ -38,8 +38,39 @@ const getMediaQueries = (ast, filter) => (
...
+}
+
+const getStyles = classNames => {
+  const styles = styleSheet
```
Comment (MicheleBertoli) [inline src/serializers/styleSheetSerializer.js]: This is not going to work with Styled Components v2, please check #14.
You should apply your changes to the AST.

**Comment by MicheleBertoli** [comment]
Any update on this @tylerferguson? Thanks!

**Comment by tylerferguson** [comment]
Hello @MicheleBertoli! Sorry I haven't revisited this lately. I got talking to @philpl last night at a JS meetup and he had the suggestion of incorporating a feature like this as an option in jest-styled-components. I thought it was interesting idea. What do you think - would it just be confusing for users if there were two quite different approaches?

**Comment by MicheleBertoli** [comment]
Thanks @tylerferguson @philpl, that's an interesting solution.
I was actually thinking about creating two separate matchers like `toMatchStyledComponentsSnapshot` (the existing one) and `toMatchVisibleStylesSnapshot` (or whatever) :)
Any idea how to cover the [edge cases](https://github.com/styled-components/jest-styled-components/pull/16#issuecomment-300068342)?

**Comment by MicheleBertoli** [comment]
Closing this due to inactivity, let's talk more about it in case you are still interested in creating a new matcher @tylerferguson. Thanks!

**Pull request closed without merging**